CryptoIntegration fails to init on WebSphere Liberty

### Describe the bug

In 19.0.0 and later versions the **PemUtils** and **CertUtils** classes fail to initialize on WebSphere Liberty with the following exception:

> Caused by: java.lang.ExceptionInInitializerError
> 	at java.lang.J9VMInternals.ensureError(J9VMInternals.java:145)
> 	at java.lang.J9VMInternals.recordInitializationFailure(J9VMInternals.java:134)
> 	at org.keycloak.adapters.KeycloakDeploymentBuilder.internalBuild(KeycloakDeploymentBuilder.java:69)
> 	at org.keycloak.adapters.KeycloakDeploymentBuilder.build(KeycloakDeploymentBuilder.java:205)
> 	at ***.iseries.api.tokenutils.TokenRetriever.<init>(TokenRetriever.java:51)
> 	at ***.iseries.api.tokenutils.TokenRetriever.<init>(TokenRetriever.java:55)
> 	at ***.iseries.tokenserver.Worker.post(Worker.java:74)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
> 	at java.lang.reflect.Method.invoke(Method.java:508)
> 	at com.ibm.ejs.container.interceptors.InterceptorProxy.invokeInterceptor(InterceptorProxy.java:200)
> 	at com.ibm.ejs.container.interceptors.InvocationContextImpl.proceed(InvocationContextImpl.java:585)
> 	at com.ibm.ws.cdi.ejb.impl.InterceptorChain.proceed(InterceptorChain.java:119)
> 	at com.ibm.ws.cdi.ejb.impl.EJBCDIInterceptorWrapper.invokeInterceptors(EJBCDIInterceptorWrapper.java:131)
> 	at com.ibm.ws.cdi.ejb.impl.EJBCDIInterceptorWrapper.postConstruct(EJBCDIInterceptorWrapper.java:67)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
> 	at java.lang.reflect.Method.invoke(Method.java:508)
> 	at com.ibm.ejs.container.interceptors.InterceptorProxy.invokeInterceptor(InterceptorProxy.java:189)
> 	at com.ibm.ejs.container.interceptors.InvocationContextImpl.proceed(InvocationContextImpl.java:577)
> 	at org.jboss.weld.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:73)
> 	at com.ibm.ws.cdi.ejb.interceptor.WeldSessionBeanInterceptorWrapper.aroundInvoke(WeldSessionBeanInterceptorWrapper.java:58)
> 	at com.ibm.ws.cdi.ejb.interceptor.WeldSessionBeanInterceptorWrapper.postConstruct(WeldSessionBeanInterceptorWrapper.java:67)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
> 	at java.lang.reflect.Method.invoke(Method.java:508)
> 	at com.ibm.ejs.container.interceptors.InterceptorProxy.invokeInterceptor(InterceptorProxy.java:189)
> 	at com.ibm.ejs.container.interceptors.InvocationContextImpl.proceed(InvocationContextImpl.java:577)
> 	at com.ibm.ejs.container.interceptors.InvocationContextImpl.doLifeCycle(InvocationContextImpl.java:326)
> 	at com.ibm.ejs.container.SingletonBeanO.callTransactionalLifecycleInterceptors(SingletonBeanO.java:221)
> 	... 24 more
> Caused by: java.lang.IllegalStateException: Not able to load any cryptoProvider with the classLoader: sun.misc.Launcher$AppClassLoader@bf56248d
> 	at org.keycloak.common.crypto.CryptoIntegration.detectProvider(CryptoIntegration.java:52)
> 	at org.keycloak.common.crypto.CryptoIntegration.init(CryptoIntegration.java:27)
> 	at org.keycloak.common.util.PemUtils.<clinit>(PemUtils.java:41)
> 	... 55 more


I believe this is because they use system classloader which is not aware of application provided classes because WebSphere uses dedicated application class loader to load it's classes.

The error occurs in **detectProvider** method of **CryptoIntegration** class (stream is empty cause of service loader not loading default implementation):

```
    private static CryptoProvider detectProvider(ClassLoader classLoader) {
        List<CryptoProvider> foundProviders = StreamSupport.stream(ServiceLoader.load(CryptoProvider.class, classLoader).spliterator(), false)
                .collect(Collectors.toList());

        if (foundProviders.isEmpty()) {
            throw new IllegalStateException("Not able to load any cryptoProvider with the classLoader: " + classLoader);
        } else if (foundProviders.size() > 1) {
            throw new IllegalStateException("Multiple crypto providers loaded with the classLoader: " + classLoader +
                    ". Make sure only one cryptoProvider available on the classpath. Available providers: " +foundProviders);
        } else {
            logger.infof("Detected crypto provider: %s", foundProviders.get(0).getClass().getName());
            return foundProviders.get(0);
        }
    }
```

Shouldn't **PemUtils** and **CertificateUtils** classes be initialized exactly the same as **KeycloakDeploymentBuilder**?

`CryptoIntegration.init(KeycloakDeploymentBuilder.class.getClassLoader());`

instead of

`    static {
        CryptoIntegration.init(ClassLoader.getSystemClassLoader());
    }`

### Version

19.0.0

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Anything else?

_No response_

I have created a sample project to reproduce the issue - https://github.com/rgala/untitled.git
Just clone it and run **mvn clean package liberty:run-server** and you should get an exception as below:

> [INFO] Caused by: java.lang.IllegalStateException: Not able to load any cryptoProvider with the classLoader: jdk.internal.loader.ClassLoaders$AppClassLoader@2437c6dc

Looks like there is also an issue with applications built using spring-boot-maven-plugin (fat jar with all dependencies):

> Exception in thread "main" java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:108)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:58)
        at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:65)
Caused by: java.lang.ExceptionInInitializerError
        at org.keycloak.adapters.KeycloakDeploymentBuilder.internalBuild(KeycloakDeploymentBuilder.java:69)
        at org.keycloak.adapters.KeycloakDeploymentBuilder.build(KeycloakDeploymentBuilder.java:205)
        at TokenRetriever.<init>(TokenRetriever.java:33)
        at TokenRetriever.<init>(TokenRetriever.java:37)
        at Main.main(Main.java:19)
        ... 8 more
Caused by: java.lang.IllegalStateException: Not able to load any cryptoProvider with the classLoader: jdk.internal.loader.ClassLoaders$AppClassLoader@6d06d69c
        at org.keycloak.common.crypto.CryptoIntegration.detectProvider(CryptoIntegration.java:52)
        at org.keycloak.common.crypto.CryptoIntegration.init(CryptoIntegration.java:27)
        at org.keycloak.common.util.PemUtils.<clinit>(PemUtils.java:41)
        ... 13 more

--- a/common/src/main/java/org/keycloak/common/util/CertificateUtils.java
+++ b/common/src/main/java/org/keycloak/common/util/CertificateUtils.java
@@ -31,7 +31,7 @@ import org.keycloak.common.crypto.CryptoIntegration;
 public class CertificateUtils {
 
     static {
-        CryptoIntegration.init(ClassLoader.getSystemClassLoader());
+        CryptoIntegration.init(CertificateUtils.class.getClassLoader());
     }
 
 

--- a/common/src/main/java/org/keycloak/common/util/PemUtils.java
+++ b/common/src/main/java/org/keycloak/common/util/PemUtils.java
@@ -38,7 +38,7 @@ public class PemUtils {
     public static final String END_CERT = "-----END CERTIFICATE-----";
 
     static {
-        CryptoIntegration.init(ClassLoader.getSystemClassLoader());
+        CryptoIntegration.init(PemUtils.class.getClassLoader());
     }
 
     /**

